public/app/plugins/datasource/loki/querySplitting.ts

```diff
@@ -0,0 +1,91 @@
+import { Subscriber, map, Observable } from 'rxjs';
+
+import { DataQueryRequest, DataQueryResponse, dateTime, TimeRange } from '@grafana/data';
+import { LoadingState } from '@grafana/schema';
+
+import { LokiDatasource } from './datasource';
+import { getRanges } from './metricTimeSplit';
+import { combineResponses, resultLimitReached } from './queryUtils';
+import { LokiQuery } from './types';
+
+function partitionTimeRange(originalTimeRange: TimeRange, intervalMs: number, resolution: number): TimeRange[] {
+  // we currently assume we are only running metric queries here.
+  // for logs-queries we will have to use a different time-range-split algorithm.
+
+  // the `step` value that will be finally sent to Loki is rougly the same as `intervalMs`,
+  // but there are some complications.
+  // we need to replicate this algo:
+  //
+  // https://github.com/grafana/grafana/blob/main/pkg/tsdb/loki/step.go#L23
+
+  const start = originalTimeRange.from.toDate().getTime();
+  const end = originalTimeRange.to.toDate().getTime();
+
+  const safeStep = Math.ceil((end - start) / 11000);
+  const step = Math.max(intervalMs * resolution, safeStep);
+
+  const ranges = getRanges(
+    start,
+    end,
+    step,
+    60 * 1000 // we go with a hardcoded 1minute for now
+  );
+
+  // if the split was not possible, go with the original range
+  if (ranges == null) {
+    return [originalTimeRange];
+  }
+
+  return ranges.map(([start, end]) => {
+    const from = dateTime(start);
+    const to = dateTime(end);
+    return {
+      from,
+      to,
+      raw: { from, to },
+    };
+  });
+}
+
+export function runPartitionedQuery(datasource: LokiDatasource, request: DataQueryRequest<LokiQuery>) {
+  let mergedResponse: DataQueryResponse | null;
+  // FIXME: the following line assumes every query has the same resolution
+  const partition = partitionTimeRange(request.range, request.intervalMs, request.targets[0].resolution ?? 1);
+  const totalRequests = partition.length;
+
+  const next = (subscriber: Subscriber<DataQueryResponse>, requestN: number) => {
+    const requestId = `${request.requestId}_${requestN}`;
+    const range = partition[requestN - 1];
+    datasource
+      .runQuery({ ...request, range, requestId })
+      .pipe(
+        // in case of an empty query, this is somehow run twice. `share()` is no workaround here as the observable is generated from `of()`.
+        map((partialResponse) => {
+          mergedResponse = combineResponses(mergedResponse, partialResponse);
+          return mergedResponse;
+        })
+      )
+      .subscribe({
+        next: (response) => {
+          if (requestN > 1 && resultLimitReached(request, response) === false) {
```

Comment:
We will need it in the follow-up PR, and we're ignoring log requests right now (it will return false for metric queries). I would like to keep it, but don't mind removing it if you think this shouldn't be here.

Reply:
No worries, was just an idea. Let's keep it there then.